aten/src/ATen/native/TensorIteratorReduce.cpp

```diff
@@ -28,21 +30,33 @@ static bool use_two_pass_reduction(TensorIterator& iter) {
 }
 
 static void two_pass_reduction(TensorIterator& iter, const loop2d_t& loop) {
-  int num_threads = at::get_max_threads();
+  int max_threads = at::get_max_threads();
 
   auto& dst = iter.tensor(0);
   auto buffer_shape = DimVector(dst.sizes());
-  buffer_shape.insert(buffer_shape.begin(), num_threads);
+  buffer_shape.insert(buffer_shape.begin(), max_threads);
   auto buffer = at::empty(buffer_shape, dst.type());
 
+  std::unique_ptr<bool[]> written(new bool[max_threads]);
+  std::fill(written.get(), written.get() + max_threads, false);
+
   at::parallel_for(0, iter.numel(), internal::GRAIN_SIZE, [&](int64_t begin, int64_t end) {
-    auto slice = buffer[at::get_thread_num()];
+    int thread_num = at::get_thread_num();
+    written[thread_num] = true;
+    auto slice = buffer[thread_num];
     slice.copy_(dst);
 
     auto sub_iter = TensorIterator::reduce_op(slice, iter.tensor(1));
     sub_iter->serial_for_each(loop, {begin, end});
   });
 
+  // fill any unwritten slices of the buffer with the identity
+  for (int thread_num = 0; thread_num < max_threads; thread_num++) {
```

Comment:
The copied slices here will be handled by the `final_reduce`, right?

Reply:
Yes. So for a sum() call buffer may start out like:

```
buffer = [nan, nan, nan, nan]  # uninitialized
```

After the parallel_for it may be something like:

```
buffer = [55, 20, nan, nan]  # partially written
```

After this for-loop:

```
buffer = [55, 20, 0, 0]  # fully written
```

And then the final reduce below, `dst` will contain 75.

In reality this is a bit of "defensive" programming. The change to `parallel_reduce()` above means that `two_pass_reduction` isn't called within a parallel region, so unless the user has more than 32,768 OpenMP threads the entire buffer will be written.

Reply:
Ah makes sense! Thanks :)